evaluator/src/main/kotlin/DependencyRule.kt

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2017-2019 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.evaluator
+
+import com.here.ort.model.Identifier
+import com.here.ort.model.LicenseFinding
+import com.here.ort.model.Package
+import com.here.ort.model.PackageLinkage
+import com.here.ort.model.PackageReference
+import com.here.ort.model.Project
+import com.here.ort.model.Scope
+
+/**
+ * A [Rule] to check a single [dependency][PackageReference].
+ */
+class DependencyRule(
+    ruleSet: RuleSet,
+    name: String,
+    pkg: Package,
+    detectedLicenses: List<LicenseFinding>,
+
+    /**
+     * The [dependency][PackageReference] to check.
+     */
+    val dependency: PackageReference,
+
+    /**
+     * The parents of this [dependency] in the dependency tree. The first entry is the root of the tree, the last entry
+     * is the direct parent.
+     */
+    val parents: List<PackageReference>,
+
+    /**
+     * The level of this [dependency] inside the dependency tree. Starts with 1 for root level entries.
+     */
+    val level: Int,
+
+    /**
+     * The [Scope] that contains this [dependency].
+     */
+    val scope: Scope,
```

Comment:
Then the rule would be applied again for the other scope, as `DependencyRule`s are applied to each single node in the tree, without any deduplication.

Reply:
Could we clarify this in the docs somehow? Maybe by saying something like

> The [Scope] of the [dependency] this rule is currently applied to.

Reply:
I have added a sentence to the `dependencyRule` function in `RuleSet`, because there is the code that decides how the `DependencyRule` is applied. Also, testing in the next weeks will show how much further refactoring is required, I don't want to make the documentation perfect already.